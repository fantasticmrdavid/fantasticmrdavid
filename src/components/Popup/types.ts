export type StateProps = {
  isActive: boolean,
  isLoading: boolean,
};

export type DispatchProps = {
  close: () => void;
};

export type ContainerProps = {
  close: (...args: any[]) => any,
  orientation: string,
  title: string,
  url: string,
};
