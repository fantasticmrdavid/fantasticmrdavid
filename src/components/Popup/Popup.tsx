import React, { memo, SyntheticEvent, useContext } from 'react';
import { LoadingContext } from 'contexts/Loading';
import {
  Container,
  Lightbox,
  Dialog,
  Close,
  Landscape,
  Portrait,
} from './popup.styles';

export type PopupProps = {
  isActive: boolean,
  onClose: (...args: any[]) => any,
  orientation: string,
  title: string,
  url: string,
};

export const Popup = memo(
  (props: PopupProps) => {
    const {
      onClose,
      isActive,
      orientation,
      title,
      url,
    } = props;
    const { getIsLoading } = useContext(LoadingContext);

    const handleClose = (e: SyntheticEvent) => {
      e.preventDefault();
      onClose();
    };

    const Image = orientation === 'landscape' ? Landscape : Portrait;

    return (
      <Container isLoading={getIsLoading()} isActive={isActive}>
        <Lightbox data-testid={"popupLightbox"} onClick={handleClose} />
        <Dialog>
          <Close data-testid={"popupClose"} onClick={handleClose}>X</Close>
          { !!url && <Image data-testid={"popupImage"} data-orientation={orientation} src={url} alt={title} /> }
        </Dialog>
      </Container>
    );
  },
);
