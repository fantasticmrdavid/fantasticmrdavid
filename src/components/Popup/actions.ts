import { Dispatch } from 'redux';
import {
  POPUP_SHOWED,
  POPUP_HIDDEN,
} from 'reducers/popup';
import {
  POPUP_LOADING_STARTED,
  POPUP_LOADING_STOPPED,
} from 'reducers/loading';

export const popupShowed = (popup: any) => (dispatch: Dispatch<any>) => {
  dispatch({ type: POPUP_LOADING_STARTED });
  dispatch({
    type: POPUP_SHOWED,
    popup,
  });
  setTimeout(() => dispatch({ type: POPUP_LOADING_STOPPED }));
};

export const popupHidden = () => ({
  type: POPUP_HIDDEN,
});
