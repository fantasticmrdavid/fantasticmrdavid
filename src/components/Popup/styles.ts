import styled from 'styled-components';
import { transparentize } from 'polished';
import theme from 'styles/theme';
import { media } from 'styles/utils';

interface ContainerProps {
  isLoading: boolean,
  isActive: boolean,
}

export const Container = styled.section<ContainerProps>`
  position: fixed;
  width: 100%;
  height: 100%;
  z-index: ${(props) => (props.isLoading || !props.isActive ? -1 : 26)};
  display: flex;
  justify-content: center;
  align-items: center;
  transition: 0.5s;
  opacity: ${(props) => (props.isLoading || !props.isActive ? 0 : 1)};
`;

export const Lightbox = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  background: ${transparentize(0.3, theme.colorBlack)};
`;

export const Dialog = styled.div`
  position: relative;
  border: ${`1px solid ${theme.colorCharcoal}`};
  box-shadow: ${theme.dropShadow};
  max-width: 90%;
  max-height: 90%;
  transition: 0.5s;
`;

export const Landscape = styled.img`
  width: 1000px;
  max-width: 90vw;
  height: auto;

  ${media.xlarge`
    width: 95vw;
    max-width: 1600px;
  `}
`;

export const Portrait = styled.img`
  height: auto;
  width: auto;
  max-height: 90vh;
  max-width: 90vw;

  ${media.small`
    height: 100vh;
    max-width: auto;
  `}
`;

export const Close = styled.div`
  position: absolute;
  top: 0px;
  right: 0px;
  padding: 0.5em;
  color: white;
  background-color: ${transparentize(0.8, theme.colorBlack)};
  font-size: 1.4rem;
  cursor: pointer;
  transition: 0.3s;
  z-index: 2;

  ${media.small`
    right: -1em;
    padding: 0;
  `};

  ${media.hover`
    &:hover {
      color: tomato;
    }
  `};
`;
