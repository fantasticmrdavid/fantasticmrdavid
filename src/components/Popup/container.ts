import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AppState } from 'reducers';
import { popupHidden } from './actions';
import Ui from './ui';

type StateProps = {
  active: boolean,
  loading: boolean,
};

type DispatchProps = {
  close: () => void;
};

const mapStateToProps = (state: AppState) => {
  return {
    ...state.popup,
    active: state.popup.active,
    loading: state.loading.popup,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>) => {
  return { close: () => dispatch(popupHidden()) };
};

export default connect<StateProps, DispatchProps, {}, AppState>(
  mapStateToProps,
  mapDispatchToProps,
)(Ui);
