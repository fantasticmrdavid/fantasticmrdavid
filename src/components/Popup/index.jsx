import React, { memo } from 'react';
import PropTypes from 'prop-types';
import * as styles from './styles';

const Popup = memo(
  (props) => {
    const {
      active,
      close,
      loading,
      orientation,
      title,
      url,
    } = props;

    const handleClose = (e) => {
      e.preventDefault();
      close(e);
    };

    const {
      Container,
      Lightbox,
      Dialog,
      Close,
      Landscape,
      Portrait,
    } = styles;

    const Image = orientation === 'landscape' ? Landscape : Portrait;

    return (
      <Container loading={loading} active={active}>
        <Lightbox onClick={handleClose} />
        <Dialog>
          <Close onClick={handleClose}>X</Close>
          { !!url && <Image src={url} alt={title} /> }
        </Dialog>
      </Container>
    );
  },
);

Popup.propTypes = {
  close: PropTypes.func.isRequired,
  title: PropTypes.string,
  url: PropTypes.string,
  orientation: PropTypes.string,
  active: PropTypes.bool,
  loading: PropTypes.bool,
};

export default Popup;
