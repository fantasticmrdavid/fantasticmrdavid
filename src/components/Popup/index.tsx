import React, { memo, SyntheticEvent } from 'react';
import * as styles from './styles';

interface Props {
  active: boolean,
  close: (...args: any[]) => any,
  loading: boolean,
  orientation: string,
  title: string,
  url: string,
}

export default memo(
  (props: Props) => {
    const {
      active,
      close,
      loading,
      orientation,
      title,
      url,
    } = props;

    const handleClose = (e: SyntheticEvent) => {
      e.preventDefault();
      close(e);
    };

    const {
      Container,
      Lightbox,
      Dialog,
      Close,
      Landscape,
      Portrait,
    } = styles;

    const Image = orientation === 'landscape' ? Landscape : Portrait;

    return (
      <Container loading={loading} active={active}>
        <Lightbox onClick={handleClose} />
        <Dialog>
          <Close onClick={handleClose}>X</Close>
          { !!url && <Image src={url} alt={title} /> }
        </Dialog>
      </Container>
    );
  },
);
