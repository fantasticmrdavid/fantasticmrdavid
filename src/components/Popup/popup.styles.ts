import styled from "styled-components";
import { transparentize } from "polished";
import theme from "styles/theme";
import { media } from "styles/utils";
import { fadeIn } from "styles/animations";

interface ContainerProps {
  $isLoading: boolean;
  $isActive: boolean;
}

export const Container = styled.section<ContainerProps>`
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  z-index: ${(props) => (props.$isLoading || !props.$isActive ? -1 : 26)};
  display: flex;
  justify-content: center;
  align-items: center;
  transition: 0.5s;
  opacity: ${(props) => (props.$isLoading || !props.$isActive ? 0 : 1)};
`;

export const Lightbox = styled.div`
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  background: ${transparentize(0.3, theme.colorBlack)};
`;

export const Dialog = styled.div`
  position: relative;
  border: ${`1px solid ${theme.colorCharcoal}`};
  box-shadow: ${theme.dropShadow};
  max-width: 90%;
  max-height: 90%;
  transition: 0.5s;
`;

export const Landscape = styled.img`
  width: auto;
  max-width: 90vw;
  height: auto;
  max-height: 90vh;
  opacity: 0;
  animation: ${fadeIn} 0.5s forwards;

  ${media.xlarge`
    max-width: 1600px;
  `}
`;

export const Portrait = styled.img`
  height: auto;
  width: auto;
  max-height: 90vh;
  max-width: 90vw;
  opacity: 0;
  animation: ${fadeIn} 0.5s forwards;

  ${media.small`
    max-width: auto;
  `}
`;

export const Close = styled.div`
  position: absolute;
  top: 0px;
  right: 0px;
  padding: 0.5em;
  color: white;
  background-color: ${transparentize(0.8, theme.colorBlack)};
  font-size: 1.4rem;
  cursor: pointer;
  transition: 0.3s;
  z-index: 2;

  ${media.small`
    right: -1em;
    padding: 0;
  `};

  ${media.hover`
    &:hover {
      color: tomato;
    }
  `};
`;
