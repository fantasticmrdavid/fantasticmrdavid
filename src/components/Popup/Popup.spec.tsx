import React from 'react';
import { render } from "@testing-library/react";
import { Popup } from 'components/Popup/Popup';

const props = {
  isActive: true,
  onClose: jest.fn(),
  loading: false,
  orientation: 'landscape',
  title: 'Popup Test',
  url: '/about',
};

describe('Component -> Popup', () => {
  it('should render the closed Popup UI without throwing an error', () => {
    expect(render(<Popup {...props} isActive={false} />)).toMatchSnapshot();
  });

  it('should render the open Popup UI without throwing an error', () => {
    expect(render(<Popup {...props} />)).toMatchSnapshot();
  });

  it('should call close() when Lightbox is clicked', () => {
    const component = render(<Popup {...props} />);
    component.getByTestId("popupLightbox").click();
    expect(props.onClose).toHaveBeenCalled();
  });

  it('should call close() when Close element is clicked', () => {
    const component = render(<Popup {...props} />);
    component.getByTestId("popupClose").click();
    expect(props.onClose).toHaveBeenCalled();
  });

  it('should render a Landscape element if a URL is provided and orientation is landscape', () => {
    const component = render(<Popup {...props} />);
    expect(component.getByTestId("popupImage")).toHaveAttribute("data-orientation", "landscape");
  });

  it('should NOT render a Portrait element if the orientation is landscape', () => {
    const component = render(<Popup {...props} />);
    expect(component.getByTestId("popupImage")).not.toHaveAttribute("data-orientation", "portrait");
  });

  it('should render a Portrait element if a URL is provided and orientation is portrait', () => {
    const component = render(<Popup {...props} orientation="portrait" />);
    expect(component.getByTestId("popupImage")).toHaveAttribute("data-orientation", "portrait");
  });

  it('should NOT render a Landscape element if the orientation is landscape', () => {
    const component = render(<Popup {...props} orientation="portrait" />);
    expect(component.getByTestId("popupImage")).not.toHaveAttribute("data-orientation", "landscape");
  });
});
