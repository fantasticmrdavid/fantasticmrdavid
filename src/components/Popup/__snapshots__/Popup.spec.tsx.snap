// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> Popup should render the closed Popup UI without throwing an error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="popupstyles__Container-sc-s4yppy-0 elHvUq"
      >
        <div
          class="popupstyles__Lightbox-sc-s4yppy-1 hRyyrn"
          data-testid="popupLightbox"
        />
        <div
          class="popupstyles__Dialog-sc-s4yppy-2 cHtwSM"
        >
          <div
            class="popupstyles__Close-sc-s4yppy-5 hhCXdn"
            data-testid="popupClose"
          >
            X
          </div>
          <img
            alt="Popup Test"
            class="popupstyles__Landscape-sc-s4yppy-3 ijffxr"
            data-orientation="landscape"
            data-testid="popupImage"
            src="/about"
          />
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="popupstyles__Container-sc-s4yppy-0 elHvUq"
    >
      <div
        class="popupstyles__Lightbox-sc-s4yppy-1 hRyyrn"
        data-testid="popupLightbox"
      />
      <div
        class="popupstyles__Dialog-sc-s4yppy-2 cHtwSM"
      >
        <div
          class="popupstyles__Close-sc-s4yppy-5 hhCXdn"
          data-testid="popupClose"
        >
          X
        </div>
        <img
          alt="Popup Test"
          class="popupstyles__Landscape-sc-s4yppy-3 ijffxr"
          data-orientation="landscape"
          data-testid="popupImage"
          src="/about"
        />
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Component -> Popup should render the open Popup UI without throwing an error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="popupstyles__Container-sc-s4yppy-0 kQNgkc"
      >
        <div
          class="popupstyles__Lightbox-sc-s4yppy-1 hRyyrn"
          data-testid="popupLightbox"
        />
        <div
          class="popupstyles__Dialog-sc-s4yppy-2 cHtwSM"
        >
          <div
            class="popupstyles__Close-sc-s4yppy-5 hhCXdn"
            data-testid="popupClose"
          >
            X
          </div>
          <img
            alt="Popup Test"
            class="popupstyles__Landscape-sc-s4yppy-3 ijffxr"
            data-orientation="landscape"
            data-testid="popupImage"
            src="/about"
          />
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="popupstyles__Container-sc-s4yppy-0 kQNgkc"
    >
      <div
        class="popupstyles__Lightbox-sc-s4yppy-1 hRyyrn"
        data-testid="popupLightbox"
      />
      <div
        class="popupstyles__Dialog-sc-s4yppy-2 cHtwSM"
      >
        <div
          class="popupstyles__Close-sc-s4yppy-5 hhCXdn"
          data-testid="popupClose"
        >
          X
        </div>
        <img
          alt="Popup Test"
          class="popupstyles__Landscape-sc-s4yppy-3 ijffxr"
          data-orientation="landscape"
          data-testid="popupImage"
          src="/about"
        />
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
